CDK-451 signal ready views

bbrownz opened the pull request:
Implementation for hdfs and hive based views, hbase is currently unimplemented.

Includes updates to interfaces for views for the basic signalReady/isReady concept

Updates were also made to the dataset repository interface to create these signals for datasets based on constraints, and to determine the last time those constraints were signaled.

MapReduce was updated to create these signals (if possible) when a write to a dataset is successful. Crunch was updated to support using these signals for WriteMode.CHECKPOINT (with a 'ready' view essentially indicating a previous success)

bbrownz commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/predicates/In.java:
```diff
@@ -116,8 +116,11 @@ public String toString() {
+    // ensure the set is sorted before calling toString on it.
+
+    Iterable<String> strings = Iterables.transform(set, new ToString<T>(schema));
+    return Joiner.on(',').join(Sets.newTreeSet(strings));
```
When making this change we came across a potential bug we wanted to chat about. After the change a test like the one below will fail 100% of the time (Smith,Jones would always be normalized to Jones,Smith). However, in the current code base this test would fail some of the time, since the ordering of the parameters isn't retained at parse time.

    String originalUri = "view:hive:person?lastName=Smith,Jones";
    String getUriResult = Datasets.load(uri).getUri().toString();
    Assert.assertEquals(originalUri, getUriResult);

rdblue replied:
I'm not sure I'm clear on what happens here. Generally, we are using a `LinkedHashMap` or `LinkedHashSet` to preserve the order. If we aren't doing that somewhere, then I think it is a bug and should be fixed.

URIs should only be normalized when needed, to avoid annoyance. I think that's what you do in this case, but I wanted to make sure.

bbrownz replied:
The parsing for In that uses a HashSet rather than a LinkedHashSet is up at [line 45](https://github.com/kite-sdk/kite/pull/348/files#diff-b835a100cf3c183c62f4afa63550e5e1R45). I can open up a new issue for that.

I think it's a stretch here to say that this code path is only when needed. This change would be hit on the normal call of View.toURI() and would result in a normalized URI. I can look at a tweak where the URI is normalized only when the caller requests it.

rdblue replied:
Thanks for opening an issue for it. I think the caller should request normalization, so that would be great.

bbrownz replied:
CDK issue ([CDK-983](https://issues.cloudera.org/browse/CDK-983)) for the predicate parsing issue.

bbrownz commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/AbstractRefinableView.java:
```diff
@@ -223,4 +223,41 @@ public URI getUri() {
+  public boolean isReady() {
+    Pair<DatasetRepository, Map<String, String>> pair =
+        Registration.lookupDatasetUri(URI.create(getUri().getRawSchemeSpecificPart()));
+    DatasetRepository repo = pair.first();
```
Currently it's fairly awkward to get back to the dataset repository that a dataset originated from. What do you guys think about adding a "getRepostiory" method to the Dataset API to it more natural to work back up the tree?

rdblue replied:
Why does the dataset delegate to the repository to determine the ready timestamp?

bbrownz replied:
Currently the ready signals are stored with the MetadataProvider. There didn't seem to be any direct way to determine the metadata provider for a Dataset (although I certainly might have missed it). My first pass exposed the metadata provider through the repository for the dataset to work against but it felt cleaner to work with the repository rather than directly with the metadata provider.

rdblue replied:
Why not have the dataset manage its own signals? I don't see a need to manage the signals at a repository/provider level because there aren't many operations that make sense. It seems like you will always want to get a view and then check if it is ready:

```java
Datasets.load("view:...").isReady();
```

ben-roling replied:
@rdblue - I presume your question here is why involve the MetadataProvider at all?  I think Ben's primary reason for involving the DatasetRepository was for access to the MetadataProvider.  I think the motivation for involving he MetadataProvider was to place the signals in a subdirectory of ".metadata".

I imagine instead they could just be placed in a ".signals" directory that sits on its own under the dataset's root directory.  If that were the case that would keep the DatasetRepository and MetadataProvider out of the picture.

Is that what you were thinking?

bbrownz commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/View.java:
```diff
@@ -128,4 +128,41 @@
+   * Note that an {@link #isEmpty() empty} view may be signaled as ready.
+   *
+   * @throws UnsupportedOperationException if the view does not support the ready
+   *          concept.
```
We went back and forth on using UnsupportedOperationException here, @ben-roling's previous pull around this concept (#100) started in this direction and worked toward another interface and custom exception for the ready concept. UnsupportedOperationException seems to more closely match the pattern established in the "deleteAll" API so I stuck with that for the initial pass, and signals feel like they could be widely supported across dataset implementations.

Now that more time has passed since that original pull request I wanted to get some thoughts on if this feels like the better direction or adding a layered interface feels more appropriate.

rdblue replied:
I like the idea of using an interface to carry these methods. If we decide later that all views should be `Signalable` (or whatever it turns out to be) then it's a binary compatible change to have `View` extend that interface. Removing those methods is not a compatible change.

Another reason for the interface is that we generally prefer compile-time problems to runtime problems. If the method always exists, then it is less likely that a caller is going to handle the `UnsupportedOperationException` case. But if you have to check for an interface and then cast, the caller has to account for views that are not signalable.

bbrownz commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/DatasetRepository.java:
```diff
@@ -181,6 +181,52 @@
   <E> Dataset<E> update(String namespace, String name, DatasetDescriptor descriptor);
 
   /**
+   * Update a dataset to indicate that a set of constraints
```
@ben-roling and I debated if adding the API at this level was correct, we explored some other options like adding a protected scope method to the AbstractDatasetRepository to allow implementations of Datasets to gain access to the MetadataProvider to call these methods there rather than adding a mirror of those methods here on the DatasetRepository.

We went with the mirrored methods for the first pass since this seemed to be a more established pattern with other methods nearly matching between the DatasetRepository and MetadataProvider (the create and update methods being examples, although their return values are different, Dataset vs DatasetDescriptors)

bbrownz commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java:
```diff
@@ -0,0 +1,229 @@
+      // does not exist it will be created
+      FileContext context = FileContext.getFileContext(conf);
+      final FSDataOutputStream os = context.create(stagedSignalPath,
+              EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE), CreateOpts.perms(FsPermission.getDefault()));
```
Looks like we might need to go back to using the filesystem rather than the file context, the travis.ci build failed on the hadoop1 profile with CreateFlag not being available

ben-roling replied:
That would be unfortunate as I don't believe there is any way to do the rename with overwrite atomically with the Filesytem API.  How long does Kite intend to continue support for hadoop1?

joey replied:
We should discuss dropping Hadoop1 support on the mailing list. I'm in favor of it, but I do think it would be considered a backwards incompatible change so it would require a version number bump.

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/TestRefinableViews.java:
```diff
@@ -819,4 +819,23 @@ public void testNotPartitioned() throws Exception {
+    Assert.assertTrue("Should be ready after signal", withSpecificTimestamp.isReady());
+
+    final RefinableView<StandardEvent> beforeSpecificTimestamp =
+        unbounded.toBefore("timestamp", instant);
```
I think this should also test instant +1, which will be in the same physical partition but shouldn't be signalled.

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/predicates/TestInToFromString.java:
```diff
@@ -76,15 +76,28 @@ public void testMultipleStringValues() {
         Predicates.in(new Utf8("a"), new Utf8("b"), new Utf8("")).toString(STRING));
     Assert.assertEquals("In#toString(STRING)",
-        "a,b,", Predicates.in("a", "b", "").toString(STRING));
+        ",a,b", Predicates.in("a", "b", "").toString(STRING));
```
I don't think these should be changed. The `In` implementation should generally preserve order unless normalized to produce a `String`. Normalization should probably be a peer of `toString(Schema)`, `toNormalizedString(Schema)`.

rdblue commented on kite-data/kite-data-crunch/src/main/java/org/kitesdk/data/crunch/DatasetTarget.java:
```diff
@@ -85,7 +85,14 @@ public boolean handleExisting(WriteMode writeMode, long lastModForSource,
+      // view isn't readiable, continue as if it wasn't ready
+    }
+    // a view exists if it isn't empty, or if it has been marked ready
+    boolean exists = !view.isEmpty() || ready;
```
Can you describe the intended behavior for checkpoint?

bbrownz replied:
This change makes a couple behavior updates. 

In WriteMode.DEFAULT a view could be considered to exist if it was empty but had been signed as ready. So in normal operation if a crunch stage (successfully) had an output view that received no data, but was marked as "ready", a subsequent run would fail with the RuntimeException in the case statement below. Before this would only have happened if the dataset/view was not empty.

In the CHECKPOINT case the same changes around what constitutes a View that exists also apply. Beyond that in checkpoint mode the pipeline would be fast-forwarded to that output point if the output view is newer than the input, and the output is signaled ready. Otherwise if the output view is overwritten if the source view has been modified more recently, or if the output view was never signaled ready. For example if an external process wrote to a view, but didn't signal it, checkpoint would remove the existing data and produce new output data.

It's worth noting too that the PR updates the "last modified" concept to consider the time a view was marked as ready. So if I produced a starting dataset at T0, validated it, and then marked it ready at T1, the view would be considered as modified at T1. (if it was never marked as ready, it will still evaluate last modified as T0)

rdblue replied:
Sounds like `signalReady` is basically the same thing as a FS "touch": it causes the view to exist if it did not already and updates the modified time. Sounds good to me. We will need to make sure the behavior is correct for both MR and Crunch.

rdblue commented:
@bbrownz this is looking good, but I have a couple general comments:
* I think the `SignalManager` should be used by the view or dataset directly rather than going through the repo and provider, though I'm interested to hear what your reasoning was for doing it this way.
* Can you say what the effects are for MR and Crunch? Should MR also be signal aware?
* I think the constraint normalization should work like `toString(Schema)`

Thanks for working on this!

bbrownz commented:
Pushed an update to address Ryan's comments above. 

Signals are now stored in a ".signals" directory with the FIleSystemDataset data itself. My original intent was that the signals metadata would go with the other metadata, but this feels as good. It also had the advantage of making the change much less drastic.

Made a couple of additions to Predicate and Constraints to have methods to request normalized toString and queryMaps.

The updates to how the change effects crunch and map/reduce are summarized above https://github.com/kite-sdk/kite/pull/348#discussion_r28354108. However, it's possible that we could make an additional update to MapReduce for the DEFAULT write mode [here](https://github.com/kite-sdk/kite/pull/348/files#diff-fba44dac726ba4b54d4fd8e2c3f5538eR479) that would support failing writes into an empty but ready view.

I'll also potentially need to make an update to the SignalManager to use the FileSystem vs FileContext APIs to support the hadoop-1 profile.

ben-roling commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/ReadiableView.java:
```diff
@@ -0,0 +1,55 @@
+package org.kitesdk.data;
+
+/**
+ * Readiable views can be used to signal to consumers that a particular view is
```
Maybe this is a bit of a nitpick, but the description seems a bit awkward to me.

"Readiable views can be used to signal..."

It almost sounds like if you want to signal that some arbitrary view X is ready then you need to find some other (readiable) view Y whose purpose is to give you the ability to signal that X is ready.

The purpose of a ReadiableView isn't to signal.  The purpose is to provide access to a subset of some Dataset.

I would suggest something like "A Readiable view is a view that may signal consumers when the underlying data is ready for consumption.  Not all View implementations provide this capability."

Furthermore, although the interface is named ReadiableView, there are no methods on the interface that actually represent a View.  As such, something could implement this interface and not be a View at all.  With the given definition I think the interface would more appropriately be named simply "Readiable".  Alternatively, the interface could extend View such that it really does represent something that is both a View and Readiable.

bbrownz commented:
I pushed a series of commits that we'll need to squash before merging. Mostly so we can look at the changes individually

* Updated the DatasetKeyOutputFormat so MR jobs in default write mode will fail if the view is empty but has been signaled as ready
* Updated SignalManager to use the FileSystem API rather than FileContext. This will fixed the hadoop-1 profile (less important, looks like that will be going away). But it will give us support for webhdfs URIs as @mkwhitacre noted on  PR #350. webhdfs doesn't have a backing implementation for the AbstractFileSystem APIs. It moves the ready signal times to the file modification timestamps, so we lose a bit of resolution, but it seems worth the tradeoff.
* Updated the ReadiableView interface to extend View as @ben-roling noted [here](https://github.com/kite-sdk/kite/pull/348#discussion-diff-28618877). This felt better than just "Readiable" since some of the wording for future implementations for covered-by and contained-by signals on the isReady API feel very View centric.

bbrownz commented:
@mkwhitacre was kind enough to kick off the travis-ci build for me and it looks like the hadoop-1 profile still isn't going through due to test failures in the mapreduce project.

I managed to track this down to issues with the LocalJobRunner around the way that it retrieves the output committer differently between tasks and the job. More details can be found on [MAPREDUCE-2350](https://issues.apache.org/jira/browse/MAPREDUCE-2350), [MAPREDUCE-4280](https://issues.apache.org/jira/browse/MAPREDUCE-4280), [MAPREDUCE-4281](https://issues.apache.org/jira/browse/MAPREDUCE-4281).  Essentially it doesn't use the committer specified by the output format for job level commit details (which is where the signal is created).

We can fix this up for the tests but it would require changing the NullOutputCommitter from a org.apache.hadoop.mapreduce.OutputCommitter to a org.apache.hadoop.mapred.OutputCommitter. Alternatively, I could add some "Assume(!Hadoop1)" statements to the tests involving the ready signals. Thoughts from anyone on which way would be better to go with?

rdblue commented:
Thanks @bbrownz, I'm going to spend some quality time reviewing this tomorrow. Let's plan on the assumption for now, considering we will probably not support Hadoop 1 for much longer.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/Constraints.java:
```diff
@@ -456,6 +456,21 @@ public String toString() {
+   * be equal in value and iteration order for any logically equivalent set of
+   * constraints.
+   */
+  public Map<String, String> toNormalizedQueryMap() {
```
Minor: this could be a bit simpler if you used a private implementation method and passed in the map container to use. Then toQueryMap could pass in a LinkedHashMap and the normalized version could use a TreeMap.

bbrownz replied:
Good call, implemented here: 855856a

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/ReadiableView.java:
```diff
@@ -0,0 +1,52 @@
+ * Readiable views may signal consumers when their underlying data is ready for
+ * consumption. Not all View implementations provide this capability.
+ */
+public interface ReadiableView<E> extends View<E> {
```
I think the original PR for this used "Signalable" instead of "Readiable", and I like that better. I see how this name is slightly more precise because the signal that is implemented is a "ready" signal. But, "readiable" isn't a commonly-used word (at least, "signalable" is less awkward to me) and I think it implies that the object can be *made ready* instead of *signalled ready*.

bbrownz replied:
Updated to Signalable here: 253cbae

I think having the future compatible change to include a ReadySignalable (when other potentially exclusive types of signals are added) makes it a good decision.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemDataset.java:
```diff
@@ -562,4 +572,35 @@ public boolean isEmpty() {
+
+    Path signalsPath = new Path(directory, SIGNALS_DIRECTORY_NAME);
+
+    SignalManager signalManager = SignalManager.create(fileSystem.getConf(), signalsPath);
```
It's a little awkward to have dataset call `signalReady` on unbounded, and then have unbounded call `signalReady` on the dataset. I think a cleaner way to do this is to have a package-local `getSignalManager` that everything can use via `dataset.getSignalManager()`. Then the dataset can lazily create the signal manager and also cache the instance.

bbrownz replied:
The change commented [here](https://github.com/kite-sdk/kite/pull/348#discussion_r29079115) removed this awkwardness. The dataset isReady/signalReady methods still call into the unbounded view, but the call back up into the dataset is removed.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemView.java:
```diff
@@ -214,6 +215,24 @@ public long getLastModified() {
+  @Override
+  public boolean isReady() {
+    long readyTimestamp = ((FileSystemDataset<E>)dataset).getReadyTimestamp(getConstraints());
+    return readyTimestamp != -1;
```
Shouldn't this be `readyTimestamp > 0`?

bbrownz replied:
It could be and would probably operate the same way, the ready timestamp is -1 (by the API in the SignalManager) when the view hasn't been signaled as ready so this feels more correct. Zero would be a valid timestamp for a ready signal exactly at the epoch, and in the same vein negative values shouldn't be possible.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemView.java:
```diff
@@ -214,6 +215,24 @@ public long getLastModified() {
+
+  @Override
+  public void signalReady() {
+    ((FileSystemDataset<E>)dataset).signalReady(getConstraints());
```
I don't think we should cast to `FileSystemDataset` here. I'd prefer to either pass the signal manager to the views, like we do with the partition listener, or find some other way, like adding getSignalManager to AbstractRefinableView.

bbrownz replied:
I tried both routes and I think injecting the SignalManager into the FileSystemView on construction (similar to partition listener) is the better route.  The AbstractRefinableView route had the issue that the SignalManager itself is an implementation, so if HBase decided to go a different route than the FileSystem storage it would be awkward at best. Going the constructor route hides any of the details of how the signals are managed.

820d9c8f

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/TestRefinableViews.java:
```diff
@@ -819,4 +820,27 @@ public void testNotPartitioned() throws Exception {
+    final ReadiableView withSpecificTimestamp =
+        (ReadiableView) unbounded.with("timestamp", instant);
+
+    Assert.assertFalse("Should not be ready initially", withSpecificTimestamp.isReady());
```
I think you can make the same assertion about unbounded.

bbrownz replied:
Updated to check the unbounded views state: a794050

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/TestRefinableViews.java:
```diff
@@ -819,4 +820,27 @@ public void testNotPartitioned() throws Exception {
+    final ReadiableView immediatelyAfterSpecificTimestamp =
+        (ReadiableView) unbounded.with("timestamp", instant+1);
+    Assert.assertFalse("Views in the same partition should not be ready",
+        immediatelyAfterSpecificTimestamp.isReady());
```
This should also validate that a view that includes the timestamp is not ready.

bbrownz replied:
Along with the changes to include a more exhaustive set of tests around the different ways to specify the view constraints I updated this to be more of boundary test around comparable values. Covering instants immediately before and after, containing ranges, and boundaries before and after the ready view (inclusive/exclusive)

a794050

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/TestRefinableViews.java:
```diff
@@ -819,4 +820,27 @@ public void testNotPartitioned() throws Exception {
   }
+
+  @Test
+  public void testReadySignals() {
```
I'd like to see a more exhaustive set of views tested by this method. Like the other tests in this file, this should make sure views backed by different constraints can be signaled correctly.

bbrownz replied:
Added some additional tests cover the different RefinableView possibilities 9d3ba86

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/filesystem/TestFileSystemDataset.java:
```diff
@@ -609,6 +612,111 @@ public void testDeleteAllWithoutPartitions() {
+        NOT_READY_TIMESTAMP, ds.getReadyTimestamp(viewConstraints));
+    long currentTime = System.currentTimeMillis();
+    // wait some time to ensure we can distinguish the current time vs the signal time
+    Thread.sleep(1000);
```
Sleeping makes tests take a much longer time. Instead, I suggest using a kind of spinlock that will wait until the condition you want is satisfied:

```java
while (currentTime == signalTime) {
  currentTime = System.currentTimeMillis();
}
```

bbrownz replied:
I added a spin lock here: 9a60125. Although it's a bit awkward. Since we're waiting on the lower resolution file system timestamps to update we don't have a great way to see when we've waited long enough other than creating more files (repeatedly signalingReady and checking the timestamp)

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/filesystem/TestFileSystemDataset.java:
```diff
@@ -609,6 +612,111 @@ public void testDeleteAllWithoutPartitions() {
+    Constraints unsignaledConstraints = new Constraints(USER_SCHEMA).
+        with("username", "joe");
+
+    ds.signalReady(viewConstraints);
```
These are good tests, but I think they should be written to the exposed API instead of using the internal implementation methods. If we remove `FileSystemDataset#signalReady(Constraints)` in the future, then these tests will break even though we still have a consistent API.

bbrownz replied:
Updating to passing the signal manager into the view on construction removed the need for the Dataset.signalReady(constraints) methods and removed the need for some of the tests at this level, the ones that remained made much more sense to use the exposed APIs for the tests. The change can be seen here: 820d9c8f

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/predicates/TestInToFromString.java:
```diff
@@ -117,6 +127,16 @@ public void testMultipleIntegerValues() {
+    // ensures that cases where the type may not be comparable directly
+    // that they will be consistently sorted
+    Assert.assertEquals("In#toString(INT)",
+        "123,45", Predicates.in(45,123).toNormalizedString(INT));
```
Minor: It would be nice to have a human-friendly sort order rather than lexicographic. Not a big deal if it's a pain.

bbrownz replied:
I think this would be possible, although a little awkward (different behaviors in the switch for the different Schema types). I think it's worth it to leave it as just lexicographic in all cases. It isn't as human friendly but it is highly consistent.

rdblue replied:
That works. Just keep in mind that we can't change it. :)

rdblue commented on kite-data/kite-data-crunch/src/main/java/org/kitesdk/data/crunch/DatasetTarget.java:
```diff
@@ -85,7 +86,12 @@ public boolean handleExisting(WriteMode writeMode, long lastModForSource,
+      ready = ((ReadiableView)view).isReady();
+    }
+    // a view exists if it isn't empty, or if it has been marked ready
+    boolean exists = !view.isEmpty() || ready;
```
Ready is a cheaper check, so that should go first. Then we can avoid accessing files in some cases.

bbrownz replied:
a84e2d8

rdblue commented on kite-data/kite-data-crunch/src/main/java/org/kitesdk/data/crunch/DatasetTarget.java:
```diff
@@ -85,7 +86,12 @@ public boolean handleExisting(WriteMode writeMode, long lastModForSource,
 
-    boolean exists = !view.isEmpty();
+    boolean ready = false;
+    if(view instanceof ReadiableView) {
```
Nit: missing a space: `if(`

rdblue replied:
Looks like this happened it a lot of places.

bbrownz replied:
This should cover all of them, 7392007. Hadn't noticed I was dropping the spaces.

rdblue commented on kite-data/kite-data-crunch/src/main/java/org/kitesdk/data/crunch/DatasetTarget.java:
```diff
@@ -99,12 +105,12 @@ public boolean handleExisting(WriteMode writeMode, long lastModForSource,
           }
-          if (ready && lastModForTarget > lastModForSource) {
+
+          if (ready && (lastModForTarget > lastModForSource)) {
```
The logic here looks wrong: if the target (derived data) was written after its source, then the checkpoint can be used regardless of whether it has been signalled. Plus the fact that the last modified time now includes when the view was signaled, shouldn't this leave out the check for ready entirely?

bbrownz replied:
I was equating the view being signaled as ready as the successful flag. So even if the target had been updated more recently (would a partial update be possible?) you would still want to replace a non-ready (unsuccessful) view.

However, if we removed the auto-signal with mapreduce I think we would definitely want to remove the ready check here, or at least include a check if we add a configuration option for auto-signal.

rdblue replied:
I think for this, the logic should still be in terms of the modified time. The ready signal updates that, so we get that the data was marked ready (after the source was modified) or the data was rewritten independent of the ready flag. I think we want that to be an or rather than requiring the flag. What if the data was written by something other than Kite and then consumed? There's no reason not to use it if its modified time is more recent.

bbrownz replied:
I'm not sure I'm following now

>I think we want that to be an or rather than requiring the flag. 

Are you suggesting that we keep the ready flag, but switch to "ready || (lastModForTarget > lastModForSource)"?

>What if the data was written by something other than Kite and then consumed? There's no reason not to use it if its modified time is more recent.

I think if something was written into a Kite target outside of Kite (highly dangerous I would think), or that something was written into the Kite view target (using Kite) but outside of MapReduce that the output would be highly suspicious. It seems like things from the input would have been missed, or else it would have been run through M/R in the first place. Usage for WriteMode.CHECKPOINT is usually that you're sure you'll achieve the same output by running the stage as you already have. The best way to ensure that is that the output was achieve by running through M/R (in which case the ready flag is present) and that the source is older than the target, which isn't necessarily guaranteed by the ready flag.

ben-roling replied:
Yeah, I think there is some confusion here.  I think ready has to be a requirement to pick back up at the checkpoint.  You can only use the checkpoint output if you know the checkpoint output is complete.  If you looked at modification time alone you could pick up incomplete output unless writes to the output directory are always done in one big atomic write, which isn't the case with the way Kite outputs are merged.

rdblue replied:
> If you looked at modification time alone you could pick up incomplete output

Isn't this handled by the fact that your driver program will run the pipeline correctly? The modification time is only used to see whether the source data has been updated more recently than derived data. There's no chance that the derived data, which is calculated by the pipeline, is in the progress of being written.

What I meant by the "or" statement was that marking something ready updates its modification time. That means it is either ready or modified recently by simply checking the modification time in this case. Does that make sense?

bbrownz replied:
That clarifies the comment.

I think what we're differing on is the semantics meaning of the signal. A modification time can be influenced by the ready signal time, but the modification time does not imply readiness. For example if the output is written non-atomically (say in a two part merge) and fails part way through the job fails and the ready signal is not written. A check on the modification time would be influenced by the first completed portion of the non-atomic write, but the job would be incorrect to use that partial output as a complete checkpoint. Treating the ready signal independently protects against that.

rdblue replied:
I think that makes sense. Since we aren't using an atomic operation to complete MR writes, we should use the signal.

In that case, it sounds like a useful thing to add always signal in the merge output format, like you had earlier. But, if the dataset implementation isn't Signalable, I think we should only consider the modification time. That way we don't break checkpoints for things that don't have signals implemented, like HBase.

bbrownz replied:
With HBase (right now anyway) the behavior would remain unchanged since it doesn't implement LastModifiedAccessor. So an HBase target is always considered to be older than any type of source and will never be reused.

Right now the implementations for Signalable and LastModifiedAccessor overlap so there aren't cases where we have one without the other. If a future Dataset implementation picked up LastModifiedAccsssor in order to use checkpoint writemode safely it seems like it would need to be fully atomic. In which case maybe Signalable isn't all that valuable from a map reduce perspective, since any view would be "ready" its nature.

I'm open to changing it, but it seems like without an "atomic" check that we could assert in code relying only on last modified feels dangerous.

rdblue replied:
That sounds fine to me.

rdblue commented on kite-data/kite-data-mapreduce/src/main/java/org/kitesdk/data/mapreduce/DatasetKeyOutputFormat.java:
```diff
@@ -462,8 +476,13 @@ public void checkOutputSpecs(JobContext jobContext) {
+        if(target instanceof ReadiableView) {
+          isReady = ((ReadiableView)target).isReady();
+        }
+        if (!target.isEmpty() || isReady) {
```
Check `isReady` first.

bbrownz replied:
a84e2d8

rdblue commented on kite-data/kite-data-mapreduce/src/main/java/org/kitesdk/data/mapreduce/DatasetKeyOutputFormat.java:
```diff
@@ -334,11 +335,20 @@ public void close(TaskAttemptContext taskAttemptContext) {
+      View<E> targetView = load(jobContext);
+
+      if(targetView instanceof ReadiableView) {
+        ((ReadiableView)targetView).signalReady();
```
What is your reasoning behind automatically signalling that the output is ready? I don't want to accumulate a lot of unused files in the signal directory. Does it make sense to add a configuration option to signal the output when it is completed?

bbrownz replied:
This was discussed somewhat on the [jira](https://issues.cloudera.org/browse/CDK-451?focusedCommentId=22245&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22245), the conclusion at the time was to settle on always signaling, without a flag or configuration option.

This ties somewhat into the comment from the SignalManager around the names of the ready files. But my thinking here is that we could compact past ready signals down into some sort of a bulk storage. Some of the future thoughts around covered-by/contained-by views would require loading all the ready views, doing that in bulk should be more efficient. 

In the short term the ready signal files from MapReduce are analogous enough to SUCCESS files that I wouldn't think most would be worried about them accumulating.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java:
```diff
@@ -0,0 +1,194 @@
+   * @return a signal manager instance, or <code>null</code> if the given
+   * directory does not exist.
+   */
+  public static SignalManager load(Configuration conf, Path signalDirectory) {
```
I think it would be a cleaner API to always return a SignalManager, and possibly just expose the constructor instead of create and load. Then the manager can handle whether or not the directory exists internally and internal libraries can use a null manager to indicate that the underlying Dataset implementation doesn't support a signal manager. That would help solve the casting problem I noted.

bbrownz replied:
I modeled this class around the [SchemaManager](https://github.com/kite-sdk/kite/blob/release-1.0.0/kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SchemaManager.java). I'm still working through the change for AbstractRefinableView.getSignalManager, I'll look at this change alongside that.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java:
```diff
@@ -0,0 +1,194 @@
+    try{
+      // create the output stream to overwrite the current contents, if the directory or file
+      // exists it will be overwritten to get a new timestamp
+      FSDataOutputStream os = rootFileSystem.create(signalPath, true);
```
I think it would be valuable to be able to signal a path that already has been by using [`setTimes`](https://hadoop.apache.org/docs/r2.2.0/api/org/apache/hadoop/fs/FileSystem.html#setTimes(org.apache.hadoop.fs.Path, long, long)) if the file already exists. Then you'd be able to use signals to reprocess if new data comes in.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java:
```diff
@@ -0,0 +1,194 @@
+        // empty, will be thrown when the signal path doesn't exist
+      }
+      return -1;
+    } catch (NumberFormatException e) {
```
Is this needed? What throws NumberFormatException?

bbrownz replied:
Good catch 8ae9928, this was around in the previous implementation of the signal manager that made use of the FileContext API so we could get finer grained signal times than the modified timestamps themselves.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java:
```diff
@@ -0,0 +1,194 @@
+      parts.add(builder.toString());
+    }
+
+    return Joiner.on('&').join(parts);
```
Lets make sure '&' is a good thing to use here and have a few tests with multiple constraints. It's a little awkward because it is a reserved character for most shells too. Is there something better we could use instead?

I'm not sure we will be needing to go from a signal to a set of constraints, so I don't really have a problem using a hash of the value just to get more reliable names. Or maybe write the constraints into the file as content?

bbrownz replied:
The '&' is safe for HDFS. Although is certainly awkward for shell usage I'm not sure how often you'd interact with these files through a shell. I added another test for signal manager with multiple constraints here: ecb9065. Additionally the '&' mirrors the query portion of the view URI so it feels like the most natural fit.

Covered-by, and contained-in views would require that we be able to convert from the ready signal files to the constraints, and not having to read the blocks to get the contents of hashed file names to get the constraints would speed up loading for that use case.

We have a stop gap solution we were thinking of using in house before we go to Oozie 4 to use the signal files themselves for our input events. This would require the file names to be predictable and simple and a hash would throw a bit of a wrench in that idea. I wouldn't add to much weight to this for deciding if we should go with hashes or not, but it is a good example of how a creative consumer might get some use out of readable file names. (in addition to easier debugging)

rdblue replied:
Ok, sounds good to me. I also really like the idea of being able to list the signals themselves as input events by looking for signal files created after some timestamp.

bbrownz commented:
I rebased this off of master and squashed all my review progress commits down for this change.

It looks like the only issue we might not have consensus on is the behavior around WriteMode.CHECKPOINT in crunch [(comment thread here)](https://github.com/kite-sdk/kite/pull/348#discussion_r29085617)

bbrownz commented:
@rdblue is there anything I can do to help move this PR along? I wasn't sure if maybe you were waiting until you get CDK-972 with pull #360 in, it looks like this branch might need a bit of rework after that goes in (although I haven't tried a merge yet). I should be able open another PR for [CDK-476](https://issues.cloudera.org/browse/CDK-476) after this one goes in.

rdblue commented:
@bbrownz, sorry about taking so long to get back to you. I was out for the past week at Strata London. I'll look over this one last time to make sure it's good to go.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/Constraints.java:
```diff
@@ -448,12 +448,26 @@ public String toString() {
       String name = entry.getKey();
       Schema fieldSchema = SchemaUtil.fieldSchema(schema, strategy, name);
-      query.put(name, Predicates.toString(entry.getValue(), fieldSchema));
+      queryMap.put(name, Predicates.toNormalizedString(entry.getValue(), fieldSchema));
```
Sorry, this is my fault. I suggested using a single method, but I forgot that we don't want to change the form of the predicate for `toQueryMap()`. This should only call `toNormalizedString` when called through the normalized accessor. Either a flag passed to the private method or reversing this change would fix it.

bbrownz replied:
I should have thought about it a bit more when I switched it over. The addition of the flag to normalize the predicate (or not) was minor enough that it still reduced the amount of code. Updated in the latest push.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemDataset.java:
```diff
@@ -463,6 +469,13 @@ public long getLastModified() {
     }
+
+    // use the ready time if it is more recent
+    long viewModified = unbounded.getLastModified();
```
This delegation will recursively check the file system, so the logic above can be removed. I think this can just delegate to `unbounded.getLastModified()`

bbrownz replied:
I think I followed, but just to be sure, I fully delegated this method to the "unbounded.getLastModified" since at this point it would be duplicating the above filesystem traversal.

rdblue replied:
Yep, that looks good to me.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemView.java:
```diff
@@ -214,6 +219,31 @@ public long getLastModified() {
     }
+
+    // if view was marked ready more recently count it as the modified time
+    if (signalManager != null) {
```
Is `signalManager` ever null?

rdblue replied:
Nevermind. I see that it is null for testing. That should be okay.

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/TestRefinableViews.java:
```diff
@@ -17,19 +17,25 @@
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Set;
+
```
Nit: Looks like unnecessary newlines snuck into the patch.

bbrownz replied:
Oops, next time around I'm definitely going to invest some more time making sure eclipse isn't quite so opinionated.

bbrownz commented:
@rdblue no worries :) I hope the conference went well. I updated the PR with the latest comments and I'll wait for the latest travis build to go through to make sure I didn't miss anything

bbrownz commented:
Looks like last nights travis build went through, I rebased the PR off the current master and updated the commit date so it appears to be more recent.

rdblue commented:
Merged! Thanks for all your work on this @bbrownz!

bbrownz commented:
It's a pleasure :smile: I'll see if I can get the PR up for the oozie portions on https://issues.cloudera.org/browse/CDK-476 today.
